w3-store.md

```diff
@@ -1,631 +1,870 @@
+  size                  int
+  origin    optional    &ContentArchive
+}
+```
 
-Similarly, a CAR can potentially contain many data items. This is true even if the CAR has only a single root CID. For example, when storing a CAR containing a nested directory structure, you could create one "upload" for the root of the directory structure, and a separate upload for a file nested inside.
+###### Store List Failure
 
-### `upload/*`
+```ipldsch
+type StoreListFailure struct {
+  message       string
+}
+```
 
-> Delegate all abilities in the `upload/*` namespace.
+## Upload Capabilities
 
-The `upload/*` capability can be delegated to a user agent, but cannot be invoked directly. Instead, it allows the audience to derive any capability in the `upload/` namespace, provided the resource URI matches the one in the `upload/*` capability delegation.
+Capabilities under `upload/*` namespace can be used to manage list of top level content entries. While not required, it is generally assumed that user content like file will be turned into [UnixFS] DAG packed and stored in space as one or more [content archive][CAR]s. The root of the DAG is then added to the upload list.
 
-The `upload/*` capability (and all capabilities in the `upload/` namespace) can be derived from a [`*`][ucan-spec-top] capability with a matching resource URI.
+### Upload Capabilities IPLD Schema
 
-### `upload/add`
+```ipldsch
+type UploadCapability union {
+  UploadAddCapability      "upload/add"
+  UploadGetCapability      "upload/get"
+  UploadListCapability     "upload/list"
+  UploadRemoveCapability   "upload/remove"
+} representation inline {
+  discriminantKey "can"
+}
+```
 
-> Add an upload to a space.
+### Upload Add
 
-`upload/add` can be invoked to register a given DAG as being contained in a given set of CARs. The resulting "upload" will be associated with the memory space identified by the DID in the `with` field.
+Authorized agent MAY invoke `upload/add` capability on the [space] subject (`with` field) to request that provider (`aud` field) include content identified by `nb.root` in the list of content entries for the space.
 
-An `upload/add` invocation requires the root CID of the DAG to register as an "upload," along with the CID of one or more CARs that contain the complete DAG. It is expected that these CARs have previously been stored in the space using [`store/add`](#storeadd); a service provider MAY return an error response if this is not the case.
+It is expected that CARs containing content are stored in the space using [`store/add`] capability. Provider MAY enforce this invariant by failing invocation or choose to succeed invocation but fail to serve the content when requested.
 
-#### Derivations
+> ⚠️ Behavior of calling `upload/add` with a same `root` and different `shards` is not specified by the protocol. w3up reference implementation allows such invocations and updates `shards` to union of all shards across invocations.
```

Comment:
maybe worth adding a note that there may exist limit of shards at implementation level?